### PR TITLE
Enable copying of nested calldata dynamic arrays to storage in evmasm pipeline

### PR DESCRIPTION
**Pull request by r0qs**
This is a follow-up PR suggested in https://github.com/argotorg/solidity/pull/15984#discussion_r2615916271

After the changes in https://github.com/argotorg/solidity/pull/15984, the evmasm pipeline now supports copying nested calldata dynamic arrays to storage by utilizing existing Yul utility functions. Previously, the old assembly implementation could not handle such cases. In https://github.com/argotorg/solidity/pull/15984, we intentionally maintained this limitation temporarily for backward compatibility.

### CODE REVIEW

File: test/libsolidity/semanticTests/array/copying/nested_array_of_structs_memory_to_storage.sol [hunk 1]
```diff
@@ -25,12 +25,16 @@ contract C {
         return s3;
     }
 }
-// ====
-// compileViaYul: true
```
Comment (r0qs) [inline test/libsolidity/semanticTests/array/copying/nested_array_of_structs_memory_to_storage.sol:29]: Note that, this test was restricted to via-IR only, but it would never actually be captured by the removed assertion. The test copies `S[][]` from memory to storage, but since the base type is an array, it enters the nested array check (i.e. `if` block) instead of the `else` block. And since it copies from memory (not calldata), the assertion passes.

Comment (r0qs) [reply test/libsolidity/semanticTests/array/copying/nested_array_of_structs_memory_to_storage.sol:29]: Although https://github.com/argotorg/solidity/pull/15984/ refactored this part of the code, the behavior appears to predate those changes. The original code already had the same if-else structure that routes `S[][]` to the nested array check instead of the struct check for this specific case. See the original code here: https://github.com/argotorg/solidity/pull/15984/changes#diff-870828b8ab9a87c930141e3bde431c7976e1ef6c848cd2473a9cf018ab1e5ec8L184

File: libsolidity/codegen/ArrayUtils.cpp
Comment (cameel) [inline libsolidity/codegen/ArrayUtils.cpp:1]: The PR needs a changelog entry.

File: test/libsolidity/semanticTests/array/copying/array_of_struct_calldata_to_storage.sol
Comment (cameel) [inline test/libsolidity/semanticTests/array/copying/array_of_struct_calldata_to_storage.sol:1]: What's up in this file? It has zero changes but it is still present in the diff.

It's also an array of structs so I think it should rather be in #16370.

File: test/libsolidity/semanticTests/array/copying/nested_dynamic_array_element_calldata_to_storage.sol [hunk 2]
```diff
@@ -26,11 +26,12 @@ contract C {
...
 // ----
 // test(uint8[][][]): 0x20, 2, 0x40, 0x60, 0, 2, 0x40, 0x80, 1, 7, 2, 8, 9
 // gas irOptimized: 137891
+// gas legacy: 140273
+// gas legacyOptimized: 138487
 // test2(uint8[][]): 0x20, 2, 0x40, 0x80, 1, 7, 2, 8, 9
 // gas irOptimized: 164249
-// gas legacyOptimized: 120228
+// gas legacy: 167430
+// gas legacyOptimized: 164772
```
Comment (cameel) [inline test/libsolidity/semanticTests/array/copying/nested_dynamic_array_element_calldata_to_storage.sol:37]: Was the replaced expectation stale? I guess #16348 was not enough to get rid of them. Would be good to be able to enforce that test that only run via Yul can't have legacy cost expectations and vice-versa.

File: test/libsolidity/semanticTests/array/copying/nested_array_of_structs_memory_to_storage.sol (hunk 1 above)
Comment (cameel) [reply test/libsolidity/semanticTests/array/copying/nested_array_of_structs_memory_to_storage.sol:29]: No, #15984 did change behavior here. Try compiling via evmasm it with 0.8.31 and 0.8.32. On 0.8.32 it passes, while on 0.8.31 it reported the `Copying of type struct C.S memory[] memory to storage is not supported in legacy` error.

Not sure how, but apparently our new condition was not entirely equivalent and it's passing through more stuff. Can you investigate how? I would definitely have expected this to fall into the `sourceBaseType->category() == Type::Category::Array` branch on 0.8.31 as it's an array of arrays and base type should be an array, but apparently it had not.

Comment (cameel) [reply test/libsolidity/semanticTests/array/copying/nested_array_of_structs_memory_to_storage.sol:29]: Also, `elements_of_nested_array_of_structs_memory_to_storage.sol` seems to have the same issue. You should remove `compileViaYul` from it as well.

File: test/libsolidity/semanticTests/array/copying/array_of_struct_calldata_to_storage.sol
Comment (r0qs) [reply test/libsolidity/semanticTests/array/copying/array_of_struct_calldata_to_storage.sol:1]: Sorry. It only changed the spacing:
```diff
-pragma abicoder               v2;
+pragma abicoder v2;
```

I was fixing one test case with an updated gas cost and decided to fix other files that had the same spacing. I should have put this in a separate commit or PR.

File: test/libsolidity/semanticTests/array/copying/nested_dynamic_array_element_calldata_to_storage.sol (hunk 2 above)
Comment (r0qs) [reply test/libsolidity/semanticTests/array/copying/nested_dynamic_array_element_calldata_to_storage.sol:37]: Yes, the legacy expectation was never being validated since the test suite it only enforces expectations for pipelines that actually run.

Should I create an issue to track the enforcement mechanism you suggested? i.e. validating that tests don't have gas expectations for pipelines that aren't enabled.

File: test/libsolidity/semanticTests/array/copying/array_of_struct_calldata_to_storage.sol
Comment (cameel) [reply test/libsolidity/semanticTests/array/copying/array_of_struct_calldata_to_storage.sol:1]: Ah. No it's fine. Now that you mention that this was a whitespace-only change I realized that I was reviewing with whitespace changes disabled and that's why it shows like this. I thought something was actually broken with the commit instead.

I was surprised because I don't remember ever seeing this. Not sure if this just never really happened (seems unlikely) or just the new broken Github UI renders this weirdly.

File: test/libsolidity/semanticTests/array/copying/nested_dynamic_array_element_calldata_to_storage.sol (hunk 2 above)
Comment (cameel) [reply test/libsolidity/semanticTests/array/copying/nested_dynamic_array_element_calldata_to_storage.sol:37]: Sure. Not sure when we'll get to this but won't hurt to have the information about it recorded somewhere more visible.

File: test/libsolidity/semanticTests/array/copying/nested_array_of_structs_memory_to_storage.sol (hunk 1 above)
Comment (r0qs) [reply test/libsolidity/semanticTests/array/copying/nested_array_of_structs_memory_to_storage.sol:29]: > No, #15984 did change behavior here. Try compiling via evmasm it with 0.8.31 and 0.8.32. On 0.8.32 it passes, while on 0.8.31 it reported the `Copying of type struct C.S memory[] memory to storage is not supported in legacy` error.

You're right. In that PR I forgot about the recursive nature of the original legacy implementation when refactoring the assertions. The new assertions only check top-level types, missing the nested cases that the old recursive structure would catch.

File: test/libsolidity/semanticTests/array/copying/nested_dynamic_array_element_calldata_to_storage.sol (hunk 2 above)
Comment (azehagowa-creator) [reply test/libsolidity/semanticTests/array/copying/nested_dynamic_array_element_calldata_to_storage.sol:37]: Approve

File: test/libsolidity/semanticTests/array/copying/nested_array_of_structs_memory_to_storage.sol (hunk 1 above)
Comment (r0qs) [reply test/libsolidity/semanticTests/array/copying/nested_array_of_structs_memory_to_storage.sol:29]: I investigated this further, and here is the current behavior:

- **Assertion 1**: (fails when): Base type is Array AND from calldata AND dynamically encoded AND base is dynamic
   - solc-0.8.31: https://github.com/argotorg/solidity/blob/fd3a22656ebe9c91a96ebd846ab7699b5f2e053c/libsolidity/codegen/ArrayUtils.cpp#L189
   - develop: https://github.com/argotorg/solidity/blob/9dc6b2e58d55d94077f663c32b85226ba898c9fd/libsolidity/codegen/ArrayUtils.cpp#L67
- **Assertion 2**: (fails when): Base type is NOT value type AND from memory OR calldata
   - solc-0.8.31: https://github.com/argotorg/solidity/blob/fd3a22656ebe9c91a96ebd846ab7699b5f2e053c/libsolidity/codegen/ArrayUtils.cpp#L227
   - develop: https://github.com/argotorg/solidity/blob/9dc6b2e58d55d94077f663c32b85226ba898c9fd/libsolidity/codegen/ArrayUtils.cpp#L80

### Test cases

| Type | 0.8.31 | 0.8.32 | Behavior |
|------|--------|--------|----------|
| [`bytes[][] calldata -> storage`](https://github.com/argotorg/solidity/blob/4d3c0f32f1e485269cd0bc8ef28b15a7f0d10d41/test/libsolidity/semanticTests/array/copying/nested_bytes_array_calldata_to_storage.sol)* | FAILS | FAILS | PASSES (removal of Assertion 1 in this PR) |
| [`bytes[][] memory -> storage`](https://github.com/argotorg/solidity/blob/4d3c0f32f1e485269cd0bc8ef28b15a7f0d10d41/test/libsolidity/semanticTests/array/copying/nested_bytes_array_memory_to_storage.sol)* | PASSES | PASSES | PASSES (None) |
| [`bytes[] calldata -> storage`](https://github.com/argotorg/solidity/blob/4d3c0f32f1e485269cd0bc8ef28b15a7f0d10d41/test/libsolidity/semanticTests/array/copying/bytes_array_calldata_to_storage.sol)* | FAILS | FAILS | PASSES (removal of Assertion 1 in this PR) |
| [`bytes[] memory -> storage`](https://github.com/argotorg/solidity/blob/develop/test/libsolidity/semanticTests/array/copying/memory_dyn_2d_bytes_to_storage.sol) | PASSES | PASSES | PASSES (None) |
| [`S[][] calldata -> storage`](https://github.com/argotorg/solidity/blob/develop/test/libsolidity/semanticTests/array/copying/nested_array_of_structs_calldata_to_storage.sol) | FAILS | FAILS | PASSES (removal of Assertion 1 in this PR) |
| [`S[][] memory -> storage`](https://github.com/argotorg/solidity/blob/develop/test/libsolidity/semanticTests/array/copying/nested_array_of_structs_memory_to_storage.sol) | FAILS | **PASSES** | **PASSES** (unintended "feature" in 0.8.32) |
| [`S[] calldata -> storage`](https://github.com/argotorg/solidity/blob/develop/test/libsolidity/semanticTests/array/copying/array_of_struct_calldata_to_storage.sol) | FAILS | FAILS | FAILS (Assertion 2) |
| [`S[] memory -> storage`](https://github.com/argotorg/solidity/blob/develop/test/libsolidity/semanticTests/array/copying/array_of_struct_memory_to_storage.sol) | FAILS | FAILS | FAILS (Assertion 2) |
| [`uint[][] calldata -> storage`](https://github.com/argotorg/solidity/blob/develop/test/libsolidity/semanticTests/array/copying/nested_dynamic_array_element_calldata_to_storage.sol) | FAILS | FAILS | PASSES (removal of Assertion 1 in this PR) |
| [`uint[][] memory -> storage`](https://github.com/argotorg/solidity/blob/develop/test/libsolidity/semanticTests/array/copying/array_nested_memory_to_storage.sol) | PASSES | PASSES | PASSES (None) |

* tests with * were added in this PR for coverage

#### 0.8.31 Behavior
```
copyArrayToStorage(S[][] memory)
|
+-- Base type of S[][] is S[] (Array)
+-- Goes into IF block (Assertion 1)
+-- Assertion 1: Not from calldata -> PASSES
+-- RECURSIVE CALL: copyArrayToStorage(S[] memory)
    |
    +-- Base type of S[] is S (Struct, not Array)
    +-- directCopy? NO (source is memory, not storage)
    +-- Is S a value type? NO (it's a struct)
    +-- Goes into ELSE block (Assertion 2)
    +-- Assertion 2: S is NOT a value type AND from memory -> FAILS
    +-- ERROR: "Copying of type S[] memory to storage not supported"
```

#### 0.8.32 Behavior
```
copyArrayToStorage(S[][] memory)
|
+-- Base type of S[][] is S[] (Array)
+-- Goes into IF block (Assertion 1)
+-- Assertion 1: Not from calldata -> PASSES
+-- NO RECURSIVE CALL anymore
+-- Calls Yul utility function directly
+-- Yul function handles everything correctly -> WORKS
```

The current problem, as far as I can tell, is that `S[][] memory -> storage` was accidentally enabled in 0.8.32. This happened because the top-level check only examines the immediate base type (`S[]`, which is an Array), without recursively verifying that `S[]` contains `S`.

Comment (r0qs) [reply test/libsolidity/semanticTests/array/copying/nested_array_of_structs_memory_to_storage.sol:29]: @cameel, how should we handle this? Should we add a retroactive changelog note about the unintended feature enabled in 0.8.32?

Comment (cameel) [reply test/libsolidity/semanticTests/array/copying/nested_array_of_structs_memory_to_storage.sol:29]: Adding a changelog entry retroactively is fine. Just not sure if this is significant enough that I'd consider it a feature worth mentioning. It's only some corner cases.

Comment (cameel) [reply test/libsolidity/semanticTests/array/copying/nested_array_of_structs_memory_to_storage.sol:29]: Technically, the fact that it was enabled early is a bug and we don't list those in the changelog (we don't know they're there), just the fixes :)

**Comment by azehagowa-creator** [comment]
approve

File: test/libsolidity/semanticTests/array/copying/bytes_array_calldata_to_storage.sol [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+contract C {
+    bytes[] s;
+
+    function test(bytes[] calldata _a) public returns (bytes[] memory) {
+        s = _a;
+        return s;
+    }
+}
+// ----
+// test(bytes[]): 0x20, 1, 0x20, 5, "hello" -> 0x20, 1, 0x20, 5, "hello"
```
Comment (clonker) [inline test/libsolidity/semanticTests/array/copying/bytes_array_calldata_to_storage.sol:12]: You could extend the test to cover empty array and array at the boundary between short/long array conversion as well as calling it with mixed lengths

```suggestion
// test(bytes[]): 0x20, 1, 0x20, 5, "hello" -> 0x20, 1, 0x20, 5, "hello"
// test(bytes[]): 0x20, 1, 0x20, 0 -> 0x20, 1, 0x20, 0
// test(bytes[]): 0x20, 1, 0x20, 31, hex"6161616161616161616161616161616161616161616161616161616161616100" -> 0x20, 1, 0x20, 31, hex"6161616161616161616161616161616161616161616161616161616161616100"
// test(bytes[]): 0x20, 1, 0x20, 32, hex"6262626262626262626262626262626262626262626262626262626262626262" -> 0x20, 1, 0x20, 32, hex"6262626262626262626262626262626262626262626262626262626262626262"
// test(bytes[]): 0x20, 1, 0x20, 33, hex"6363636363636363636363636363636363636363636363636363636363636363", hex"6300000000000000000000000000000000000000000000000000000000000000" -> 0x20, 1, 0x20, 33, hex"6363636363636363636363636363636363636363636363636363636363636363", hex"6300000000000000000000000000000000000000000000000000000000000000"
// test(bytes[]): 0x20, 4, 0x80, 0xa0, 0xe0, 0x120, 0, 31, hex"6161616161616161616161616161616161616161616161616161616161616100", 32, hex"6262626262626262626262626262626262626262626262626262626262626262", 33, hex"6363636363636363636363636363636363636363636363636363636363636363", hex"6300000000000000000000000000000000000000000000000000000000000000" -> 0x20, 4, 0x80, 0xa0, 0xe0, 0x120, 0, 31, hex"6161616161616161616161616161616161616161616161616161616161616100", 32, hex"6262626262626262626262626262626262626262626262626262626262626262", 33, hex"6363636363636363636363636363636363636363636363636363636363636363", hex"6300000000000000000000000000000000000000000000000000000000000000"
```

**Review by clonker — COMMENTED** [review]
i think the test coverage could be a little bit better around the short/long boundary, otherwise the changes look good to me

**Comment by azehagowa-creator** [comment]
Approve and merge
